src/main/java/org/owasp/esapi/codecs/JSONCodec.java

```diff
@@ -0,0 +1,242 @@
+ */
+public class JSONCodec extends AbstractIntegerCodec {
+
+
+    /**
+     * {@inheritDoc}
+     *
+     * Escape special characters in JSON strings.
+     *
+     * encodeCharacter will escape the characters Backspace (\b), Form Feed (\f),
+     * Carriage Return (\r), Line Feed (\n), Tab (\t), Double Quote (") and Backslash (\).
+     * If the character is a control character (U+0000 through U+001f), then it will be
+     * Unicode encoded (\u0000 through \u001f). If the character is not special or in the
+     * user supplied immune list, then the character is returned unescaped. If the
+     * character is null then an empty string is returned.
+     *
+     * WARNING: This method will silently discard an invalid code point according to
+     * the result of {@code Character.isValidCodePoint( int )} method.
+     *
+     * @param immune character array of whitelist characters which should not be encoded
+     * @param c the character to encode if not in the immune list
+     * @return encoded character if the character is special, and the character otherwise.
+     */
+    public String encodeCharacter( char[] immune, Character c ) {
+        if ( c == null ) {
+            return "";
+        }
+
+        return encodeCharacter(immune, charToCodepoint( c ));
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Escape special characters in JSON strings.
+     *
+     * encodeCharacter will escape the characters Backspace (\b), Form Feed (\f),
+     * Carriage Return (\r), Line Feed (\n), Tab (\t), Double Quote (") and Backslash (\).
+     * If the character is a control character (U+0000 through U+001f), then it will be
+     * Unicode encoded (\u0000 through \u001f). If the character is not special or in the
+     * user supplied immune list, then the character is returned unescaped. If the
+     * character is null then an empty string is returned.
+     *
+     * WARNING: This method will silently discard an invalid code point according to
+     * the result of {@code Character.isValidCodePoint( int )} method.
+     *
+     * @param immune character array of whitelist characters which should not be encoded
+     * @param c the character to encode if not in the immune list
+     * @return encoded character if the character is special, and the character otherwise.
+     */
+    public String encodeCharacter( char[] immune, int codePoint )
+        throws IllegalArgumentException {
+
+        // Per the docs for HTMLEntityCodec: "WARNING: This method will silently discard
+        // invalid code points per the call to Character.isValidCodePoint( int ) method.
+        // WARNING!! Character based Codecs will silently transform code points that are
+        // not legal UTF code points into garbage data as they will cast them to chars.
+        if ( Character.isValidCodePoint( codePoint ) == false ) {
+            // throw new IllegalArgumentException( "Invalid codepoint '" + String.format("0x%04X", codePoint) + "'." );
```

Comment:
This appears to be unused, or commented for debugging purposes.  Please remove or adjust as intended.

Reply:
Also, it's problematic using "UTF" language with the Character class.  Char is from 0-65535.  The whole point of creating the AbstractIntegerCodec was to extend to the full UTF range and traditionally this library pretty much used char/Character as shorthand for Latin-1/ASCII as the developers saw fit.  If this needs to be kept that comment needs to be more reflective of this.

Reply:
Note that part of this problem is from Java's own use of "Character" to check codepoints.  At any rate it would be better to state that any character whose codepoint is > 0xffff is guaranteed to have an indeterminate value.  Use AbstractIntegerCodec if you want to handle higher values.

Reply:
Thanks @xeno6696.

I added you as a collaborator at https://github.com/noloader/esapi-java-legacy. Feel free to update the comments as you see fit.

Reply:
@xeno6696, tell me what you'd like this to say and I will update the baseline. I'd like to get this PR merged so I can pull it into the taglib branch so that can also be merged cleanly.

Reply:
My apologies for the delay, had family in town and also managed to get sick too.  

After some thought I would change it to:

```
// WARNING!! Character based Codecs will only handle the byte range of 0-65535 (0x0-0xffff).  
// Passing any data represented by a higher numerical value will result in a downcast thus
// destroying the original data with undefined results.  

```

Reply:
Sounds good. I lifted the original warning from HTMLEncoder, so the change should be made in several places.

Reply:
I intend to apply this update before the end of the week, unless someone else beats me to it.

Using the text identified by @xeno6696, in all locations identified by @noloader

Reply:
Upon further investigation, there are only 2 places in the baseline that use the block identified by @xeno6696 
```
        // WARNING!! Character based Codecs will silently transform code points that are
        // not legal UTF code points into garbage data as they will cast them to chars.
```

This occurs in the javadoc for `AbstractCodec.encode`
```
  /**
     * {@inheritDoc}
     * </p><p>
     * <b>WARNING!!</b>  {@code Character} based {@code Codec}s will silently transform code points that are not
     * legal UTF code points into garbage data as they will cast them to {@code char}s.
     * Also, if you are implementing an {@code Integer} based codec, these will be silently discarded
     * based on the return from {@code Character.isValidCodePoint( int )}.  This is the preferred
     * behavior moving forward.
     */
    @Override
    public String encode(char[] immune, String input) {
```
and in the `JSONCodec.encodeCharacter` inline comments identified by this review.

The other listed locations do reference the 'silent discard' behavior; however, the only concern being identified is the terminology used around the description of "invalid UTF 8 code points"

For this reason, I will only be updating the two baseline locations:
- Javadoc for `AbstractCodec.encode`
- Comment section in `JSONCodec.encodeCharacter`